### PR TITLE
QF-3714 In Projects admin page, add a owner type filter - by user or by organization

### PR DESCRIPTION
**Pull request by dddpt**
"owner__type" for the filter would suffice functionality wise.
But the filter title would then be "By type" in the UI, which is semantically unclear, hence the SimpleListFilter subclass.

Screenshot from 2024-01-29 11-08-41

### CODE REVIEW

**Comment by dddpt** [comment]
Quick question, can a team be a project owner?
If not, I will remove it from the options.

**Comment by suricactus** [comment]
No, a team cannot own a project. Can you please attach a screenshot on PRs that are screenshotable?

File: docker-app/qfieldcloud/core/admin.py [hunk 1]
```diff
@@ -568,6 +568,27 @@ class ProjectFilesWidget(widgets.Input):
...
+            (User.Type.PERSON, "Person"),
+            (User.Type.ORGANIZATION, "Organization"),
+            (User.Type.TEAM, "Team"),
+        ]
+
+    def queryset(self, request, queryset):
+        if self.value():
+            return queryset.filter(owner__type=self.value())
+        else:
+            return queryset
```
Comment (faebebin) [inline docker-app/qfieldcloud/core/admin.py:589]: Here I would prefer the same style as in other SimpleListFilters
 
```
       ...
        value = self.value()

        if value is None:
            return queryset
            
         ... rest of logic
```

File: docker-app/qfieldcloud/core/admin.py [hunk 2]
```diff
@@ -568,6 +568,27 @@ class ProjectFilesWidget(widgets.Input):
...
+    # which is semantically unclear, hence the SimpleListFilter subclass
+    title = "owner type"
+    parameter_name = "owner_type"
+
+    def lookups(self, request, model_admin):
+        return [
+            (User.Type.PERSON, "Person"),
+            (User.Type.ORGANIZATION, "Organization"),
+            (User.Type.TEAM, "Team"),
+        ]
```
Comment (faebebin) [inline docker-app/qfieldcloud/core/admin.py:583]:
```suggestion
        return (
            (User.Type.PERSON, _("Person")),
            (User.Type.ORGANIZATION, _("Organization")),
        )
```

File: docker-app/qfieldcloud/core/admin.py [hunk 3]
```diff
@@ -568,6 +568,27 @@ class ProjectFilesWidget(widgets.Input):
     template_name = "admin/project_files_widget.html"
 
 
+class OwnerTypeFilter(admin.SimpleListFilter):
+    # "owner__type" for the filter would suffice,
+    # But the filter title would then be "By type" in the UI,
+    # which is semantically unclear, hence the SimpleListFilter subclass
+    title = "owner type"
```
Comment (faebebin) [inline docker-app/qfieldcloud/core/admin.py:575]:
```suggestion
    title = _("owner type")
```

File: docker-app/qfieldcloud/core/admin.py (hunk 2 above)
Comment (faebebin) [reply docker-app/qfieldcloud/core/admin.py:583]: No, owner can only be Person or Organization:

In the core/models.py you can find:
```
class Project(models.Model):
    ...
    owner = models.ForeignKey(
        User,
        on_delete=models.CASCADE,
        related_name="projects",
        limit_choices_to=models.Q(type__in=[User.Type.PERSON, User.Type.ORGANIZATION]),
        help_text=_(
            "The project owner can be either you or any of the organization you are member of."
        ),
    )
    ...
```

**Review by faebebin — CHANGES_REQUESTED** [review]
I added a couple of improvements.

Hint: try to find existing snippets and copy paste and modify them. I actually just compared to an exisiting SimpleListFilter :)

Thanks, great work sofar!

File: docker-app/qfieldcloud/core/admin.py [hunk 4]
```diff
@@ -568,6 +568,27 @@ class ProjectFilesWidget(widgets.Input):
     template_name = "admin/project_files_widget.html"
 
 
+class OwnerTypeFilter(admin.SimpleListFilter):
+    # "owner__type" for the filter would suffice,
+    # But the filter title would then be "By type" in the UI,
+    # which is semantically unclear, hence the SimpleListFilter subclass
```
Comment (suricactus) [inline docker-app/qfieldcloud/core/admin.py:574]: I think comment is not required, as we already have done this in other `SimpleListFilter` instances. It's expected behaviour from Django.

File: docker-app/qfieldcloud/core/admin.py (hunk 1 above)
Comment (dddpt) [reply docker-app/qfieldcloud/core/admin.py]: done

File: docker-app/qfieldcloud/core/admin.py (hunk 2 above)
Comment (dddpt) [reply docker-app/qfieldcloud/core/admin.py]: done

File: docker-app/qfieldcloud/core/admin.py [hunk 5]
```diff
@@ -568,6 +568,20 @@ class ProjectFilesWidget(widgets.Input):
...
+    parameter_name = "owner_type"
+
+    def lookups(self, request, model_admin):
+        return [(User.Type.PERSON, "Person"), (User.Type.ORGANIZATION, "Organization")]
+
+    def queryset(self, request, queryset):
+        value = self.value()
+        if value is None:
+            return queryset
+        return queryset.filter(owner__type=value)
```
Comment (suricactus) [inline docker-app/qfieldcloud/core/admin.py:582]:
```suggestion
        value = self.value()

        if value is None:
            return queryset

        return queryset.filter(owner__type=value)
```
Slight nitpicking, we usually put some air around the `if`/`for`/`try/except` etc blocks :)

**Comment by suricactus** [comment]
> No, a team cannot own a project. Can you please attach a screenshot on PRs that are screenshotable?

Very small style adjusting comment. A valid screenshot in the first message is still required before we can merge this.

**Comment by dddpt** [comment]
done ;-)

**Review by suricactus — APPROVED** [review]
Please do no merge yet.

**Comment by suricactus** [comment]
Ready to merge!

**Pull request merged**